Add CMake config-file package to picojson installation

eaaltonen opened the pull request:
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

eaaltonen commented:
@duncanmmacleod: this PR is most of the way there, but I encountered a problem specifying the C std. lib and C++ compilers required for CMake's compiler sanity check.

As I see, these were specified in `jwt-cpp-feedstock` with
```
requirements:
  build:
    - {{ stdlib('c') }}  # [unix]
    - {{ compiler('cxx') }}  # [unix]
```

However, when running the build in a Docker container with `build-locally.py` and selecting `1. linux_64_` when prompted, the build results in an error
```
WARNING: failed to get install actions, retrying.  exception was: Unsatisfiable dependencies for platform linux-64: {MatchSpec("c_linux-64")}
Encountered problems while solving:
  - nothing provides requested c_linux-64

Could not solve for environment specs
The following package could not be installed
└─ c_linux-64 does not exist (perhaps a typo or a missing channel)
```

Since you are also on `compilers-feedstock`, perhaps you can advise on this..?

duncanmmacleod commented:
@eaaltonen, I think you need to regenerate the feedstock using conda-smithy to update the build configuration. See https://conda-forge.org/docs/maintainer/updating_pkgs/#rerendering-feedstocks.

Commit:
MNT: Re-rendered with conda-build 25.1.2, conda-smithy 3.47.0, and conda-forge-pinning 2025.03.18.12.47.30

Commit:
patch picojson to install a CMake config package

Patch picojson to use CMake for the build and installation and to also
install a CMake config-file package. The config-file package allows
downstream CMake builds to:

* check the version of the picojson header library
* link targets with the `picojson::picojson` interface library, which
    * determines the include directory path automatically
    * propagates the dependency and include directory path forward

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found some lint.

Here's what I've got...


For **recipe/meta.yaml**:

 * ❌ Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [31, 32, 33]


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/14522298083](https://github.com/conda-forge/conda-forge-webservices/actions/runs/14522298083). Examine the logs at this URL for more detail.</sub>

eaaltonen commented:
@duncanmmacleod : I finally found some time to test also building on windows, following https://conda-forge.org/docs/maintainer/knowledge_base/#simple-cmake-based-bldbat
and manually

1. downloading
2. patching
3. running build steps in Developer Command Prompt

I moved the resulting folder to a different location and ran CMake configure using the `picojson` package from that location.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

eaaltonen commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 25.4.2, conda-smithy 3.47.2, and conda-forge-pinning 2025.04.28.17.55.25

eaaltonen commented:
@duncanmmacleod : the error
```
(base) %SRC_DIR%>set "CONDA_SHLVL="   && 

(base) %SRC_DIR%>conda activate "%PREFIX%" 
The syntax of the command is incorrect.
```

Seems very strange. Is there a way to download `conda_test_env_vars.bat` from the pipeline and see what was actually written there?

jaimergp commented on recipe/meta.yaml:
```diff
@@ -8,19 +8,34 @@ package:
+    - test -f ${PREFIX}/lib/cmake/picojson/picojson-config-version.cmake  # [unix]
+    - test -f ${PREFIX}/lib/cmake/picojson/picojson-config.cmake  # [unix]
     - if not exist %LIBRARY_INC%\picojson\picojson.h exit 1  # [win]
+    - if not exist %LIBRARY_LIB%\cmake\picojson\picojson-targets.cmake  # [win]
+    - if not exist %LIBRARY_LIB%\cmake\picojson\picojson-config-version.cmake  # [win]
+    - if not exist %LIBRARY_LIB%\cmake\picojson\picojson-config.cmake  # [win]
```
```suggestion
    - if not exist %LIBRARY_LIB%\cmake\picojson\picojson-targets.cmake exit 1  # [win]
    - if not exist %LIBRARY_LIB%\cmake\picojson\picojson-config-version.cmake exit 1  # [win]
    - if not exist %LIBRARY_LIB%\cmake\picojson\picojson-config.cmake exit 1  # [win]
```

eaaltonen replied:
Thank you. Now I'm wondering why it worked on the earlier push...

eaaltonen commented:
@duncanmmacleod : this PR is waiting for your review.
If you are otherwise satisfied, please also let me know if I should include the [version bump commit](https://github.com/conda-forge/picojson-feedstock/commit/99c0700bd8712289899b158cefd29c7b140f4d8a) already in this PR.

duncanmmacleod commented on recipe/meta.yaml:
```diff
@@ -8,19 +8,34 @@ package:
+    # package is header-only
+    - {{ stdlib('c') }}
+    - {{ compiler('cxx') }}
   number: 0
```
This needs to be incremented to indicate a new build.

```suggestion
  number: 1
```

duncanmmacleod reviewed (changes requested):
> @duncanmmacleod : this PR is waiting for your review.
> If you are otherwise satisfied, please also let me know if I should include the [version bump commit](https://github.com/conda-forge/picojson-feedstock/commit/99c0700bd8712289899b158cefd29c7b140f4d8a) already in this PR.

I'm not really sure what the version commit means. If the version commit is designed just to indicate that a post-release change has been made, that is the purpose of the `build:number` key in the `meta.yaml` file - I think just bumping that is sufficient.

Sorry if I've misunderstood.

eaaltonen commented:
> > @duncanmmacleod : this PR is waiting for your review.
> > If you are otherwise satisfied, please also let me know if I should include the [version bump commit](https://github.com/conda-forge/picojson-feedstock/commit/99c0700bd8712289899b158cefd29c7b140f4d8a) already in this PR.
> 
> I'm not really sure what the version commit means. If the version commit is designed just to indicate that a post-release change has been made, that is the purpose of the `build:number` key in the `meta.yaml` file - I think just bumping that is sufficient.

The version commit was so that we would get a version that we could require in a dependant recipe's (_e.g. jwt-cpp_)  `requirements` section, a line such as:
```
- picojson >=1.3.0.post1
```

Can the `build:number` be somehow used in dependant recipe's `requirements` section?

duncanmmacleod commented:
> Can the build:number be somehow used in dependant recipe's requirements section?

No, I don't think so, but the solver should automatically select the higher build number it as long as it solves. I'm not sure how conda handles `.postN` version suffices anyway.

Commit:
Rebuild with CMake config-file packages included

eaaltonen commented:
> No, I don't think so, but the solver should automatically select the higher build number it as long as it solves. I'm not sure how conda handles .postN version suffices anyway.

Out of curiosity, I checked what I could find.

The tests suggest that conda fully supports {PEP 440](https://peps.python.org/pep-0440/)
https://github.com/conda/conda/blob/main/tests/models/test_version.py#L160

There even seems to be some sort of support for specifying a specific build number
https://github.com/conda/conda/blob/main/conda/models/match_spec.py#L543-L549

duncanmmacleod reviewed (approved):
LGTM, thanks @eaaltonen.

duncanmmacleod commented:
@eaaltonen, please confirm that you are finished with this PR and that I should merge it.

eaaltonen commented:
@duncanmmacleod : I can confirm that I think this PR is ready to be merged.
